components/main-chef-wrapper/main.go

```diff
@@ -0,0 +1,119 @@
+	// CLI to run that same binary on any 'chef generate XYZ' execution.
+	case "generate":
+		// TODO @afiune design a better way to isolate single sub-commands that we want to
+		// replace/improve, for instance we might not want to rewrite the entire 'chef generate'
+		// sub-command but a section of it.
+		//
+		// Example of a sub-command overwrite
+		if len(os.Args) > 2 && os.Args[2] == "build-cookbook" {
+			fmt.Println("The sub-command 'chef generate build-cookbook' has been deprecated.")
+			os.Exit(0)
+		}
+		fallthrough
+
+	default:
+		// When we land in the default case where we run the old 'chef' cli binary,
+		// we need to send the sub-command as well as all the arguments.
+		allArgs = append([]string{subCommand}, allArgs...)
+		cmd = exec.Command("chef-cli", allArgs...)
+	}
+
+	debugLog(fmt.Sprintf("Bin: %s\nArgs: %v", cmd.Path, allArgs))
+
+	cmd.Env = os.Environ()
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+
+	// TODO @afiune handle the errors in a way better manner
+	if err := cmd.Run(); err != nil {
+		if exitError, ok := err.(*exec.ExitError); ok {
+			os.Exit(exitError.ExitCode())
+		}
+		// @afiune This might only be needed when doing debugging
+		// because if we got here it means we have a different error
+		// other than a 'ExitError' so print
+		debugLog(fmt.Sprintf("Unexpected Error: \n%s", err))
+		os.Exit(7)
+	}
+}
+
+func usage() {
+	// TODO @afiune add actual usage, this might only list top level sub-commands
+	// we should avoid to add specific options per sub-command
+	fmt.Printf(`Usage:
+    chef -h/--help
+    chef -v/--version
+    chef command [arguments...] [options...]
+
+Available Commands:
+    exec                    Runs the command in context of the embedded ruby
+    env                     Prints environment variables used by Chef Workstation
+    gem                     Runs the 'gem' command in context of the embedded Ruby
+    generate                Generate a new repository, cookbook, or other component
+    shell-init              Initialize your shell to use Chef Workstation as your primary Ruby
+    install                 Install cookbooks from a Policyfile and generate a locked cookbook set
+    update                  Updates a Policyfile.lock.json with latest run_list and cookbooks
+    push                    Push a local policy lock to a policy group on the Chef Infra Server
+    push-archive            Push a policy archive to a policy group on the Chef Infra Server
+    show-policy             Show policyfile objects on the Chef Infra Server
+    diff                    Generate an itemized diff of two Policyfile lock documents
+    export                  Export a policy lock as a Chef Infra Zero code repo
+    clean-policy-revisions  Delete unused policy revisions on the Chef Infra Server
+    clean-policy-cookbooks  Delete unused policyfile cookbooks on the Chef Infra Server
+    delete-policy-group     Delete a policy group on the Chef Infra Server
```

Comment:
An interesting experiment in the golang world would be setting up all these policy commands to exist under a "policy" group, while still providing backwards compatibility. This would let us clear up the tire fire of a a flat namespace we have while still allowing the existing commands to work:

``chef policy install`` would be the same as the legacy ``chef install`` from the chef-cli ruby gem.

Reply:
Absolutely! If this gets in, we will have full control over a number, if not all our tools in the Chef Ecosystem. 

To that front, we need to have a good design to do such deprecations, new features, and more thought this main top-level wrapper.